### ISSUE TITLE
Incorporate Views AJAX History

### ISSUE THREAD

**Issue by herbdool** [backdrop/backdrop-issues#7075]
## Description of the need

Incorporate [views_ajax_history](https://github.com/backdrop-contrib/views_ajax_history) into core. It will allow the View to be bookmarked, navigated, copied and pasted even when AJAX is used to filter.

The module now uses the widely-supported [History API](https://developer.mozilla.org/en-US/docs/Web/API/History_API) browser feature to update the URL query parameters.

**Comment by quicksketch** [comment]
This would be great! I agree this would be a very good core candidate.

**Comment by herbdool** [comment]
Happy enough with the PR to share it now. Please test.

**Comment by quicksketch** [comment]
@herbdool I took a stab at modifying `ajax_view.js` to not override the global `Backdrop.ajax.prototype` methods and instead override the `ajax` object instances. PR filed at https://github.com/backdrop/backdrop/pull/5254

Also some small fixes like missing semicolons and strict type checking.

A quick check and it seems to work as before.

A small thing I noticed that is probably an issue with Views AJAX History as it exists in contrib: it does not account for clicking on table headers where the sort order is changed. That could be a follow-up.

**Comment by herbdool** [comment]
@quicksketch thanks for updating that. I've tested and reviewed your code changes. It works for me.

I found clicking on table headers for order and sort work for me actually.

**Comment by quicksketch** [comment]
> I found clicking on table headers for order and sort work for me actually.

Huh. Are you sure the entire page just didn't reload? They still work, they're just not AJAX requests.

**Comment by quicksketch** [comment]
I think this is pretty good. I would love some additional feedback from others. Any reviews would be appreciated.

@olafgrabienski maybe you could take a look?

**Comment by argiepiano** [comment]
I've done a bit of testing (no code review). I used the Content view that comes in core. The AJAX history works great, but I did notice that, when you use the pager, the site is adding a duplicate `admin/content` after the query question mark (the `/` is being encoded as `%2F`). See image:

<img width="1333" height="854" alt="Image" src="https://github.com/user-attachments/assets/93634706-d43b-456a-93cb-c75d2a0fa3e7" />

This is because the actual path for this view (`admin/content`) is different from the base path declared in Page Settings in the View (`admin/content/node`). This happens to all Views that provide a default tab for the UI. 

This doesn't seem to interfere with the navigation (since it's after the question mark), but it does look like a bug with the AJAX pager.

**Comment by argiepiano** [comment]
Side note: the table header sorting does reload the page even when AJAX enabled.

**Comment by argiepiano** [comment]
The issue I pointed above with the pager path is due to the fact that the Content view base path  as defined in Page Settings in the View is `/admin/content/node`, but because this path is the default local task for the Content screens, the actual path ends up being `/admin/content` when you click the menu item in the admin bar to see the content view.

Line 198 of `ajax_view.js` uses the content of `view_base_path` (in this case,  `/admin/content/node`) to replace it in the dynamically updated path. But the problem here is that, if you opened the view by clicking on the menu bar, the path  is `admin/content`, and therefore the `replace` in line 198 fails to work, leading to the duplicate path.

Instead of using `view_base_path` I believe we should use `view_path`, which is populated by extracting the path from `$_GET['q']` in `template_preprocess_views_view()`, and will most likely get the desired result (basically removing the duplicate portion in the path before dynamically updating it in the URL bar of the browser).

**Comment by olafgrabienski** [comment]
> `@olafgrabienski` maybe you could take a look?

As @argiepiano already had a look, I guess I don't have to add anything at the moment but I'm open for testing ideas.

**Comment by herbdool** [comment]
@argiepiano ready for testing/review again.

**Comment by herbdool** [comment]
We should also test whether we can still disable ajax history and still have it work like the original ajax functionality.

**Comment by herbdool** [comment]
I fixed the toggle for enabling ajax_history.

I think the table header sorting works via ajax in *some* cases. In a custom view it is always working on any column. On `admin/content` it only works on the title column.

**Comment by herbdool** [comment]
Actually the table header sorting issue pre-dates this PR. It may have something to do with the custom classes put on the field labels (e.g. `priority-low` etc).

Update: it happens when the "Default classes" option is unselected for a field.

**Comment by argiepiano** [comment]
It's been a packed day. I'll try to take a look tomorrow.

**Comment by quicksketch** [comment]
> Update: it happens when the "Default classes" option is unselected for a field.

Maybe a `data-` attribute should be added on the table sort links that is independent of the "default classes" option. Then the data attribute should be checked in the `view_ajax.js` instead of the class. That would avoid breaking any existing behavior and follow a better-practice of using `data-` for JS behaviors instead of classes.

I think it would be okay to modify the non-overrideable `tablesort_header()` function to add such an attribute. i.e. `data-order` and `data-sort`. That would make it reliable for Views and any other contrib/custom solutions, like a JS-behavior to sort a table entirely by JavaScript.

**Comment by herbdool** [comment]
> Maybe a data- attribute should be added on the table sort links...

@quicksketch as you mentioned before, this can be in a follow up, correct? It is a pre-existing problem for ajax in views.

**Comment by quicksketch** [comment]
> @quicksketch as you mentioned before, this can be in a follow up, correct? It is a pre-existing problem for ajax in views.

Yes, let's not hold up the whole feature for a relatively small existing issue.

An observation in the current PR. New views default to this option being off, but existing views when `update.php` is run seem to get the option turned on if they previously were AJAX-enabled.

Should this feature be off in both scenarios (new and existing AJAX views)? Or would it make sense to enable it by default when AJAX is enabled on a view?

**Comment by quicksketch** [comment]
> Should this feature be off in both scenarios (new and existing AJAX views)? Or would it make sense to enable it by default when AJAX is enabled on a view?

@herbdool mentioned in the PR that the intention is to turn on AJAX History in the update hook _if_ Views AJAX History module was enabled, since it globally turns on that behavior.

I updated the PR to take into account this default on updates. I left the AJAX History option off by default on new views.

I also added `views_ajax_history` to `backdrop_merged_modules()` to notify users that the module is no longer needed. The module is disabled on running update.php.

**Comment by quicksketch** [comment]
> Yes, let's not hold up the whole feature for a relatively small existing issue.

I pushed a commit to fix the table sorting when using non-default views markup. I added the `data-sort` and `data-order` attributes as described earlier. Turns out that Views doesn't use `tablesort_header()`, but I added it there also for consistency.

Thank you @argiepiano and @docwilmot for the code reviews. This is ready for more testing/feedback.

**Comment by herbdool** [comment]
@quicksketch I reviewed and tested your changes. They work for me! Turns out the data-order, data-sort changes were pretty small. And the update hook looks better.

**Comment by argiepiano** [comment]
Sorry I fell out of the face of the earth. It's finals week, and I'm slammed... it seems that you guys have thoroughly tested this! Thank you

**Comment by quicksketch** [comment]
Merged into 1.x for 1.34.0. Awesome! Thank you @herbdool for making the issue and PR! Thank you @argiepiano and @olafgrabienski for jumping in here!

**Comment by indigoxela** [comment]
Not sure if that even matters, but *all* views get saved in the newly added update hook, although the function comment `Enable Ajax History for all Views using AJAX` sort of suggests, only relevant ones will get touched.

That `$config->save();` in views_update_1009() happens without condition, on all views.
I just noticed, because I wanted to check, which of those JSON files got updated, and all were. On a dev site, where views_ajax_history was definitely **not** installed at all.

By intention or an oversight @herbdool ?

**Comment by quicksketch** [comment]
@indigoxela Yeah, a best practice in update hooks is to only save the config file if something changes.

**Comment by herbdool** [comment]
We can fix that here https://github.com/backdrop/backdrop-issues/issues/7114
